api/src/main/java/org/eclipse/microprofile/config/spi/ConfigSource.java

```diff
@@ -198,4 +201,103 @@ default int getOrdinal() {
      * @return the name of the configuration source
      */
     String getName();
+
+    /**
+     * Type of this config source. For backward compatibility returns {@link Type#PROPERTY} by default.
+     *
+     * @return source type
+     */
+    default Type getType() {
+        return Type.PROPERTY;
+    }
+
+    /**
+     * Whether the content of this config source can change externally.
+     * For backward compatibility returns {@code true} by default.
+     * <p>
+     * This information is required when the type is other than {@link Type#PROPERTY}.
+     * <p>
+     * Examples:
+     * <ul>
+     *     <li>Environment variables are not mutable</li>
+     *     <li>System properties may be considered mutable</li>
+     *     <li>Classpath resource from a jar file is not mutable</li>
+     *     <li>File based config source may be considered mutable</li>
+     * </ul>
+     *
+     * @return whether this is a mutable config source
+     */
+    default boolean isMutable() {
+        return true;
+    }
+
+    /**
+     * If this source {@link #getType()} is {@link Type#TREE}, this method is used to load the
+     * tree of the data.
+     *
+     * @return node content or empty optional if the underlying config source does not exist
+     */
+    default Optional<ConfigNode> load() {
+        return Optional.empty();
+    }
+
+    /**
+     * If this source {@link #getType()} is {@link Type#LAZY}, this method is used to load each
+     * node.
+     *
+     * @return config node or empty optional if the node does not exist, or this is not a lazy
+     *  config source
+     */
+    default Optional<ConfigNode> load(String key) {
+        return Optional.empty();
+    }
+
+    /**
+     * If this config source is {@link #isMutable()} and its shape changes, it should notify
+     * the configuration implementation of such a change.
+     * Once the change listener is registered by configuration implementation, this source can
+     * start checking for changes. If any change occurs, the source should invoke the runnable.
+     *
+     * @param changeRunnable runnable to call when this config source changes
+     */
+    default void changeListener(Runnable changeRunnable) {
+    }
+
+    /**
+     * Type of the config source.
+     */
+    enum Type {
+        /**
+         * This is for backward compatibility - request each property separately.
```

Comment:
Should this be `FLAT` to state all properties are visible with a flat structure?

Reply:
The naming could be anything, I just wanted to have an option that is backward compatible. If `FLAT` sounds better, no problem.